### PR TITLE
Add pitch fly mode

### PR DESCRIPTION
**Pull request by Gael-de-Sailly**
This adds a new, optional fly mode, in which the movement follows the exact direction the player is pointing at (including pitch), instead of moving horizontally and vertically separately.
It works exactly like the normal fly mode, except that all directions (forward, backward, up, down, left, right) are relative to the player's yaw *and* pitch, instead of only yaw. It can be enabled and disabled using a key (by default L).

[Demonstration on YouTube](https://www.youtube.com/watch?v=3sW_tyB2oYQ)

I spent some time testing it yesterday, it did not take me much time to handle it well, and it clearly makes easier to do some difficult moves, like flying in a cave. The moves are also smoother, I feel like in an airplane.

This PR may not be ready for merging, as there remains some weirdness (the player slightly deviates from its trajectory at the beginning and the end of a move).
Also I would like to find a better name than "pitch fly" for this feature (do you have any idea?).
I'm open to discussion about the choices I had to make (like the toggle key, the vector rotation that is performed at the end of the calculation instead of the beginning, etc.), I would also like to have some feedback about the concept itself before doing more polishing if needed.

### CODE REVIEW

**Comment by paramat** [comment]
Good flying in that video!

**Comment by JurajVajda** [comment]
This could be theoretically used for wings and temp fly privs :)

**Comment by benrob0329** [comment]
It'd be nice if fly mode(s) could have a "gravity" value added, so that mods can make it more or less "glidey".

**Comment by rubenwardy** [comment]
This sort of thing should be controllable by games/mods, really. It would be nice to have a unified API for it - to move players into different fly/fast/zoom/etc modes

**Comment by DS-Minetest** [comment]
I don't know how good this pitch fly mode is for really playing. It might be confusing and annoying. Other games have it like this, too (eg. Sauerbraten). But with a separate key, like it is now, I like it. Videos look really nice with this flying mode.

However should this be added to the engine? Won't SSCSM make it able for mods to do something like this? Does that speak against this? How basic is this feature? 

More client side features make Minetest nicer because it adds more content to play around with.

=> :+1: 

What about other fly modes?
Like this one: https://www.youtube.com/watch?v=CeTjR5-IabY

**Comment by paramat** [comment]
> This could be theoretically used for wings and temp fly privs :)

> It'd be nice if fly mode(s) could have a "gravity" value added, so that mods can make it more or less "glidey" for wings, or a plane, etc.

I can understand it's tempting to think these things when looking at this feature, but fly mode is very much the wrong place to add wings/glider dynamics or simulating a flying vehicle, as fly mode is by nature completely free movement, an admin/creative thing, without the restrictions you would need for wings/gliders.

Fly mode is also player movement, not vehicle movement. Wings/glider mods will want to apply specific and widely-varying behaviours, which is best done (and can only be done) using vehicle mod Lua code, instead of attempting to apply vehicle type behaviours to engine player movement (which would end up very limited anyway).

To be clear though, i do certainly support improving vehicles by having a sort of 'vehicle API' that implements vehicles like the player is implemented now. I'm just saying this PR is not the place for attempting vehicle features, it's just an alternative fly mode control.
I do like this feature, it will be good for dev work, i spend a lot of time flying over landscapes.

**Comment by Gael-de-Sailly** [comment]
I agree with @paramat. Although I would love to see natively supported aircraft-like physics, this is not what I intended to do here.

**Comment by ClobberXD** [comment]
Splendid! This is a great idea, and it just feels much more natural than the current implementation of free_move. :+1:

**Comment by stujones11** [comment]
> Fly mode is also player movement, not vehicle movement.

This could almost be made into a vehicle/plane style mod by substituting the player model with that of the vehicle and attaching a player model to it. I IRC, someone already did something like this in a mod.

**Comment by Gael-de-Sailly** [comment]
Fixed the acceleration glitch, and rebased.
I now consider my PR to be ready for review.

File: builtin/settingtypes.txt [hunk 1]
```diff
@@ -204,6 +207,10 @@ keymap_rangeselect (Range select key) key KEY_KEY_R
 #    See http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3
 keymap_freemove (Fly key) key KEY_KEY_K
 
+#    Key for toggling pitch fly mode.
+#    See http://irrlicht.sourceforge.net/docu/namespaceirr.html#a54da2a0e231901735e3da1b0edf72eb3
+keymap_freemove (Pitch fly key) key KEY_KEY_L
```
Comment (ClobberXD) [inline builtin/settingtypes.txt]: `keymap_pitchfly` :)

File: src/localplayer.h [hunk 2]
```diff
@@ -146,8 +146,7 @@ class LocalPlayer : public Player
 	void setZoomFOV(float zoom_fov) { m_zoom_fov = zoom_fov; }
 
 private:
-	void accelerateHorizontal(const v3f &target_speed, const f32 max_increase);
-	void accelerateVertical(const v3f &target_speed, const f32 max_increase);
+	void accelerate(const v3f &target_speed, const f32 max_increase_H, const f32 max_increase_V, const bool use_pitch);
```
Comment (ClobberXD) [inline src/localplayer.h]: Line looks too long, add a line-break maybe?

File: builtin/settingtypes.txt (hunk 1 above)
Comment (Gael-de-Sailly) [reply builtin/settingtypes.txt]: of course :)

File: src/localplayer.cpp [hunk 3]
```diff
@@ -729,38 +725,47 @@ v3f LocalPlayer::getEyeOffset() const
...
+	flat_speed.rotateXZBy(-yaw);
+	if (use_pitch)
+		flat_speed.rotateYZBy(-pitch);
+
+	v3f d_wanted = target_speed - flat_speed;
+	v3f d = v3f(0,0,0);
+
+	// Then compare the horizontal and vertical components with the wanted speed
+	if (max_increase_H > 0)
+	{
```
Comment (SmallJoker) [inline src/localplayer.cpp]: Wrong bracket placement.
`if (max_increase_H > 0) {`

File: src/localplayer.cpp [hunk 4]
```diff
@@ -729,38 +725,47 @@ v3f LocalPlayer::getEyeOffset() const
...
+			d += d_wanted_H;
+	}
 
-// Vertical acceleration (Y), X and Z directions are ignored
-void LocalPlayer::accelerateVertical(const v3f &target_speed, const f32 max_increase)
-{
-	if (max_increase == 0)
-		return;
+	if (max_increase_V > 0)
+	{
```
Comment (SmallJoker) [inline src/localplayer.cpp]: Bracket placement.

File: builtin/settingtypes.txt (hunk 1 above)
Comment (SmallJoker) [reply builtin/settingtypes.txt]: Would you please be so nice and update the README file as well?

File: src/localplayer.cpp [hunk 5]
```diff
@@ -689,10 +688,7 @@ void LocalPlayer::applyControl(float dtime, Environment *env)
 		slip_factor = getSlipFactor(env, speedH);
 
 	// Accelerate to target speed with maximum increment
-	accelerateHorizontal(speedH * physics_override_speed,
-			incH * physics_override_speed * slip_factor);
-	accelerateVertical(speedV * physics_override_speed,
-			incV * physics_override_speed);
+	accelerate((speedH + speedV) * physics_override_speed, incH * physics_override_speed * slip_factor, incV * physics_override_speed, pitch_fly);
```
Comment (SmallJoker) [inline src/localplayer.cpp]: Needs line breaks

File: src/localplayer.cpp [hunk 6]
```diff
@@ -729,38 +725,47 @@ v3f LocalPlayer::getEyeOffset() const
 	return v3f(0, BS * eye_height, 0);
 }
 
-// Horizontal acceleration (X and Z), Y direction is ignored
-void LocalPlayer::accelerateHorizontal(const v3f &target_speed,
-	const f32 max_increase)
+// 3D acceleration
+void LocalPlayer::accelerate(const v3f &target_speed, const f32 max_increase_H, const f32 max_increase_V, const bool use_pitch)
```
Comment (SmallJoker) [inline src/localplayer.cpp]: Needs line break

File: src/localplayer.cpp [hunk 7]
```diff
@@ -485,9 +485,7 @@ void LocalPlayer::applyControl(float dtime, Environment *env)
 
 	PlayerSettings &player_settings = getPlayerSettings();
 
-	v3f move_direction = v3f(0,0,1);
-	move_direction.rotateXZBy(getYaw());
-
+	// All vectors are relative to the player's yaw (and pitch if pitch fly mode enabled), and will be rotated at the end
```
Comment (SmallJoker) [inline src/localplayer.cpp]: Needs line break  (> 80 characters)

File: src/localplayer.cpp [hunk 8]
```diff
@@ -586,31 +587,31 @@ void LocalPlayer::applyControl(float dtime, Environment *env)
...
-		speedH += move_direction.crossProduct(v3f(0,1,0));
+		speedH += v3f(-1,0,0);
 	}
 	if (control.right) {
-		speedH += move_direction.crossProduct(v3f(0,-1,0));
+		speedH += v3f(1,0,0);
 	}
 	if (!control.left && !control.right) {
-		speedH -= move_direction.crossProduct(v3f(0,1,0)) *
+		speedH -= v3f(1,0,0) *
```
Comment (SmallJoker) [inline src/localplayer.cpp]: Just to make sure: Is this line correct? Looking at L608 (and the old code), shouldn't this be
`speedH -= v3f(-1,0,0) * ....` or `speedH += v3f(1,0,0) * ....`?

Comment (Gael-de-Sailly) [reply src/localplayer.cpp]: You are right. I did a mistake in my cross product. But this needs testing and I don't have the material for that.

**Comment by Gael-de-Sailly** [comment]
Could anyone test it with the joystick?

**Comment by lhofhansl** [comment]
Tried it (without joystick). Looks good.

**Comment by lhofhansl** [comment]
I agree with @paramat. Let's keep this simple and focused on what it tries to do (follow-pitch-fly-mode).
For more usages we should propose new PRs.

**Comment by Gael-de-Sailly** [comment]
Solving conflict with merged #7228

**Comment by lhofhansl** [comment]
Can we merge this?

**Comment by lhofhansl** [comment]
What is this waiting for?
@paramat @rubenwardy anything we need to @Gael-de-Sailly ?

**Comment by ClobberXD** [comment]
PR needs rebasing due to client-side files being refactored in #7902

**Comment by lhofhansl** [comment]
Once I get another approval I'll merge.

**Comment by SmallJoker** [comment]
I did not test the joystick control, but I suppose it will work as expected.

@lhofhansl Assuming that the approval is your own. Just in case you added it for someone else two gave :+1: above.

**Comment by lhofhansl** [comment]
@SmallJoker yeah, that was my approval.

**Comment by lhofhansl** [comment]
I'm happy to rebase at merge-time, it's just that some files have moved.

**Comment by Gael-de-Sailly** [comment]
I have one concern about it: I've seen several players saying that this feature was annoying them because when it was enabled, they weren't able to disable it. I suggest either adding the key setting in the GUI (which I totally forget to do and I apologize), or disable the key by default, like autoforward or cinematic mode (would make sense too because most players won't use it a lot).

**Comment by paramat** [comment]
Why wouldn't they be able to disable it?
It's been added to the in-game key change menu now.
It's too useful to leave the key binding empty. I find autoforward more useful than pitch move but other core devs disapprove of binding that by default, which is weird.
When new controls are added, it always annoys some players who don't know about it and enable it accidently, but that's not a reason to not add new controls, because we never would.
We could add a post in the news forum to warn about the new control.

**Comment by Gael-de-Sailly** [comment]
> Why wouldn't they be able to disable it? Do you mean when it's accidently enabled?

Yes it's about that

> It's been added to the in-game key change menu now.

Just noticed it, nice :)

**Pull request merged**